### PR TITLE
muvm: Replace socat for interactive mode.

### PR DESCRIPTION
**Pull request by WhatAmISupposedToPutHere**
Socat's tty mode has issues with window resizing
as it does not pass through SIGWINCH. Also add
a tty-less interactive mode, for usage from scripts and similar.

### CODE REVIEW

File: crates/muvm/src/bin/muvm.rs [hunk 1]
```diff
@@ -27,6 +21,12 @@ use rustix::process::{
     geteuid, getgid, getrlimit, getuid, sched_setaffinity, setrlimit, CpuSet, Resource,
 };
 use serde::{Deserialize, Serialize};
+use std::env;
```
Comment (teohhanhui) [inline crates/muvm/src/bin/muvm.rs:24]: CS: `std` imports should be a separate group on top.

File: crates/muvm/src/launch.rs [hunk 2]
```diff
@@ -56,64 +57,76 @@ impl Display for LaunchError {
...
-    Err(anyhow!("Ran out of ports."))
-}
-
-fn escape_for_socat(s: String) -> String {
-    let mut ret = String::with_capacity(s.len());
-    for c in s.chars() {
-        match c {
-            ':' | ',' | '!' | '"' | '\'' | '\\' | '(' | '[' | '{' => {
-                ret.push('\\');
+        let path = socket_dir.join(format!("port-{}.lock", port));
```
Comment (teohhanhui) [inline crates/muvm/src/launch.rs:65]:
```suggestion
        let path = socket_dir.join(format!("port-{port}.lock"));
```

File: crates/muvm/src/launch.rs [hunk 3]
```diff
@@ -56,64 +57,76 @@ impl Display for LaunchError {
...
+                lock_file
+            } else {
+                let lock_file = File::options()
+                    .write(true)
+                    .read(true)
+                    .open(path)
+                    .context("Failed to open lock file")?;
+                let ret = flock(&lock_file, FlockOperation::NonBlockingLockExclusive);
+                if ret.is_err() {
+                    continue;
```
Comment (teohhanhui) [inline crates/muvm/src/launch.rs:80]:
```suggestion
                if flock(&lock_file, FlockOperation::NonBlockingLockExclusive).is_err() {
                    continue;
```

File: crates/muvm/src/launch.rs [hunk 4]
```diff
@@ -56,64 +57,76 @@ impl Display for LaunchError {
...
-    ];
-    command = "socat".into();
-    request_launch(server_port, cookie, command, command_args, env)?;
-    socat.wait()?;
-    Ok(())
+    let run_path = env::var("XDG_RUNTIME_DIR")
+        .map_err(|e| anyhow!("unable to get XDG_RUNTIME_DIR: {:?}", e))?;
+    let socket_dir = Path::new(&run_path).join("krun/socket");
+    let (_lock, vsock_port) = acquire_socket_lock()?;
+    let path = socket_dir.join(format!("port-{}", vsock_port));
```
Comment (teohhanhui) [inline crates/muvm/src/launch.rs:107]:
```suggestion
    let path = socket_dir.join(format!("port-{vsock_port}"));
```

File: crates/muvm/src/launch.rs [hunk 5]
```diff
@@ -56,64 +57,76 @@ impl Display for LaunchError {
...
+    _ = unlink(&path);
+    let listener = UnixListener::bind(path)?;
```
Comment (teohhanhui) [inline crates/muvm/src/launch.rs:109]: Add context here?

File: crates/muvm/src/server/worker.rs [hunk 6]
```diff
@@ -1,13 +1,24 @@
+use anyhow::{anyhow, Context, Result};
+use log::{debug, error};
+use nix::errno::Errno;
+use nix::sys::epoll::{Epoll, EpollCreateFlags, EpollEvent, EpollFlags, EpollTimeout};
+use nix::sys::socket::{connect, socket, AddressFamily, SockFlag, SockType, VsockAddr};
+use nix::unistd::{pipe, setsid};
+use rustix::process::ioctl_tiocsctty;
+use rustix::pty::{ptsname, unlockpt};
+use rustix::termios::{tcsetwinsize, Winsize};
 use std::collections::HashMap;
```
Comment (teohhanhui) [inline crates/muvm/src/server/worker.rs:10]: CS: `std` imports should be a separate group on top.

File: crates/muvm/src/bin/muvm.rs (hunk 1 above)
Comment (teohhanhui) [inline crates/muvm/src/bin/muvm.rs:24]: CS: `std` imports should be a separate group on top.

File: crates/muvm/src/server/worker.rs [hunk 7]
```diff
@@ -227,5 +287,146 @@ async fn handle_connection(
...
+        }
+    })
+}
+
+fn run_io_guest(
+    stdin: OwnedFd,
+    stdout: OwnedFd,
+    stderr: Option<OwnedFd>,
+    stop_pipe: OwnedFd,
+    vsock: u32,
```
Comment (teohhanhui) [inline crates/muvm/src/server/worker.rs:326]:
```suggestion
    vsock_port: u32,
```

File: crates/muvm/src/server/worker.rs [hunk 8]
```diff
@@ -182,6 +203,8 @@ async fn handle_connection(
         command,
         command_args,
         env,
+        interactive_vsock,
```
Comment (teohhanhui) [inline crates/muvm/src/server/worker.rs:206]:
```suggestion
        interactive_vsock_port,
```

File: crates/muvm/src/server/worker.rs [hunk 9]
```diff
@@ -227,5 +287,146 @@ async fn handle_connection(
...
+    let mut stdout = File::from(stdout);
+    let mut stderr = stderr.map(File::from);
+    let mut stop_pipe = File::from(stop_pipe);
+    let vsock_fd = socket(
+        AddressFamily::Vsock,
+        SockType::Stream,
+        SockFlag::empty(),
+        None,
+    )?;
+    connect(vsock_fd.as_raw_fd(), &VsockAddr::new(2, vsock))?;
```
Comment (teohhanhui) [inline crates/muvm/src/server/worker.rs:338]:
```suggestion
    connect(vsock_fd.as_raw_fd(), &VsockAddr::new(nix::libc::VMADDR_CID_HOST, vsock))?;
```

**Review by teohhanhui — APPROVED** [review]

**Review by slp — APPROVED** [review]

**Pull request merged**